packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+    }),
+    // fsm.events.willExitState.subscribe((s) => {
+    //   log("Exiting state", s);
+    // }),
+    // fsm.events.didEnterState.subscribe((s) => {
+    //   log("Entering state", s);
+    // }),
+  ];
+  return () => {
+    for (const unsub of unsubs) {
+      unsub();
+    }
+  };
+}
+
+function defineConnectivityEvents(fsm: FSM<Context, Event, State>): {
+  statusDidChange: Observable<PublicConnectionStatus>;
+  didConnect: Observable<void>;
+  didDisconnect: Observable<void>;
+} {
+  // Emitted whenever a new WebSocket connection attempt suceeds
+  const statusDidChange = makeEventSource<PublicConnectionStatus>();
+  const didConnect = makeEventSource<void>();
+  const didDisconnect = makeEventSource<void>();
+
+  let oldPublicStatus: PublicConnectionStatus | null = null;
+
+  fsm.events.didEnterState.subscribe((newState) => {
+    const newPublicStatus = toPublicConnectionStatus(newState);
+    statusDidChange.notify(newPublicStatus);
+
+    if (oldPublicStatus === "open" && newPublicStatus !== "open") {
+      didDisconnect.notify();
+    } else if (oldPublicStatus !== "open" && newPublicStatus === "open") {
+      didConnect.notify();
+    }
+    oldPublicStatus = newPublicStatus;
+  });
+
+  return {
+    statusDidChange: statusDidChange.observable,
+    didConnect: didConnect.observable,
+    didDisconnect: didDisconnect.observable,
+  };
+}
+
+function createStateMachine<T extends BaseAuthResult>(delegates: Delegates<T>) {
+  // Create observable event sources, which this machine will call into when
+  // specific events happen
+  const onMessage = makeEventSource<IWebSocketMessageEvent>();
+
+  // Emitted whenever the server deliberately closes the connection for
+  // a specific Liveblocks reason
+  const onLiveblocksError = makeEventSource<LiveblocksError>();
+
+  const initialContext: Context & { token: T | null } = {
+    token: null,
+    socket: null,
+
+    // Bumped to the next "tier" every time a connection attempt fails (no matter
+    // whether this is for the authentication server or the websocket server).
+    // Reset every time a connection succeeded.
+    backoffDelay: LOW_DELAY,
+    // numRetries: 0,
+  };
+
+  const fsm = new FSM<Context, Event, State>(initialContext)
+    .addState("@idle.initial")
+    .addState("@idle.failed")
+    .addState("@auth.busy")
+    .addState("@auth.backoff")
+    .addState("@connecting.busy")
+    .addState("@connecting.backoff")
+    .addState("@ok.connected")
+    .addState("@ok.awaiting-pong");
+
+  //
+  // Configure events that can happen from anywhere
+  //
+  // It's always possible to explicitly get a .reconnect() or .disconnect()
+  // from the user.
+  //
+  fsm.addTransitions("*", {
+    RECONNECT: {
+      target: "@auth.backoff",
+      assign: increaseBackoffDelay,
+    },
+
+    DISCONNECT: "@idle.initial",
+  });
+
+  //
+  // Configure the @idle.* states
+  //
+  fsm.addTransitions("@idle.*", {
```

Comment:
oh right, I forgot about the wildcard part of the names, I guess that makes using enums a bit harder.